File: frontend/src/api/services/author.service.ts
import { Author, NewAuthor } from "../models/Author";
import { Page, Pagination } from "../models/Page";
import { GetManyByPagination, Save, Delete } from '@controller/AuthorController'

export async function GetAuthors(page: Page = { page: 0, size: 25 }): Promise<Pagination<Array<Author>>> {
    const response = await GetManyByPagination(page.filter ?? '', { Page: page.page, Size: page.size }, [{ Key: "id", Dir: -1 }]);

    return {
        count: response.Count,
        data: response.Data.map((a) => ({ id: a.Id, name: a.Name })),
    }
}

export async function SaveAuthor(newAuthor: NewAuthor): Promise<Author | null> {
    const response = await Save(newAuthor.name);

    return response ? { id: response.Id, name: response.Name } : null;
}

export async function DeleteAuthor(authorId: number): Promise<boolean> {
    return await Delete(authorId);
}

File: src/service/author.service.go
package service

import (
	"yotudo/src/database/errors"
	"yotudo/src/database/repository"
	"yotudo/src/lib/logger"
	"yotudo/src/model"
)

type AuthorService struct{}

var GlobalAuthorService *AuthorService = nil

func (c *AuthorService) GetManyByPagination(filter string, page *model.Page, sort []model.Sort) *model.Pagination[[]model.Author] {
	authors, totalCount := repository.GlobalAuthorRepository.FindByPage(filter, page, sort)

	return &model.Pagination[[]model.Author]{
		Data:  authors,
		Count: totalCount,
	}
}

func (c *AuthorService) Save(newAuthorName string) (*model.Author, error) {
	return repository.GlobalAuthorRepository.SaveOne(newAuthorName)
}

func (c *AuthorService) Delete(id int64) (bool, error) {
	if repository.GlobalAuthorRepository.IsReferencingToMusic(id) {
		logger.Warning("Unable to delete Author, because it was used in a music, or contributor records")

		return false, errors.ErrUnableToDelete
	}

	return repository.GlobalAuthorRepository.DeleteOne(id), nil
}
